README.md
# ci-runtimes

This command-line tool shows Github actions runtimes in a table format.

## Status

This is my personal practice project in Clojure and this tool is work-in-progress.

## Installation

- Clone this repo
- Install `leiningen`
- Create a file `config.edn` in project root and add contents:
    ```
    {:github-token "ghp_your_token"}
    ```
    where ghp_your_token is your personal Github token generated in https://github.com/settings/tokens There create a token with no scopes selected. This is to allow more requests to the github API.
- See Usage for usage instructions

## Usage

Options: `lein run <owner> <repository> "<CI run name>"`

### Example

```
❯ lein run metosin reitit testsuite
Get workflow "testsuite" of metosin/reitit ...
Got workflow id: 5883743
Get jobs of the workflow...

|      :date | :duration |           :line | :runs |
|------------+-----------+-----------------+-------|
| 2021-02-17 |       184 |     ########### |     5 |
| 2021-02-19 |       212 |   ############# |     2 |
| 2021-02-20 |       113 |         ####### |     2 |
| 2021-02-21 |       217 |   ############# |     1 |
| 2021-02-26 |        56 |            #### |     7 |
| 2021-02-27 |         4 |               # |     1 |
| 2021-03-02 |        97 |          ###### |     5 |
| 2021-03-03 |         4 |               # |     2 |
| 2021-03-05 |       125 |        ######## |     1 |
| 2021-03-11 |       187 |     ########### |    13 |
| 2021-03-12 |        73 |           ##### |     2 |
| 2021-04-09 |         4 |               # |     4 |
| 2021-04-12 |         4 |               # |     2 |
| 2021-04-22 |        91 |          ###### |     3 |
| 2021-04-23 |       137 |        ######## |     6 |
| 2021-04-30 |       101 |          ###### |     5 |
| 2021-05-26 |       135 |        ######## |     2 |
| 2021-06-21 |       189 |     ########### |     2 |
| 2021-06-24 |       232 |  ############## |     1 |
| 2021-07-27 |       241 | ############### |     1 |
| 2021-08-03 |       156 |       ######### |    13 |
| 2021-08-05 |       108 |         ####### |     2 |
| 2021-08-30 |         4 |               # |     1 |
| 2021-09-07 |       113 |         ####### |     2 |

Average duration of the workflow runs is 124.63529 seconds
```

## How this works

The inner workings of this tool.

1. Get repository workflows via https://api.github.com//repos/{owner}/{repo}/actions/workflows (https://docs.github.com/en/rest/reference/actions#list-repository-workflows)
1. From the result find workflow ID using its name
1. Get workflow's runs via https://api.github.com/repos/{owner}/{repo}/actions/workflows/{workflow_id}/runs (https://docs.github.com/en/rest/reference/actions#list-workflow-runs)
1. From the result find many job_urls: https://api.github.com/repos/{owner}/{repo}/actions/runs/{run_id}/jobs (https://docs.github.com/en/rest/reference/actions#list-jobs-for-a-workflow-run)
1. For each job_url
    1. Get job's started_at and completed_at strings
    1. Parse a Date object from those
    1. Calculate duration (as seconds) between started_at and completed_at
    1. Store the date (as YYYY-MM-DD string) and duration into a collection
1. For all stored dates and durations group the collection by date
1. For each date calculate average duration
1. For each date calculate the relative duration for illustration purposes (short duration get only a short line and long duration gets a longer line)
1. Print the results for each date

## Limitations

- Gets only the "first" job in from a run which has many jobs.

## License

Copyright © 2021 Kimmo Ahola

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

## [0.1.0] - 2021-09-15

Initial release.

[Unreleased]: https://sourcehost.site/your-name/recipes/compare/0.1.0...HEAD


test/ci_runtimes/core_test.clj
(ns ci-runtimes.core-test
  (:require [clojure.test :refer :all]
            [ci-runtimes.core :refer :all]
            [ci-runtimes.github-api :as github-api]))

(deftest a-test
  (testing "workflows-url"
    (is (= (github-api/workflows-url "some-owner" "awesome-repo")
           "https://api.github.com/repos/some-owner/awesome-repo/actions/workflows"))))


src/ci_runtimes/github_api.clj
(ns ci-runtimes.github-api
  (:require [clj-http.client :as client]
            [clojure.edn :as edn]))

(def config
  (edn/read-string (slurp "config.edn")))

(def github-token (:github-token config))

;; https://github.com/settings/tokens -> no scopes needed
(defn get-url [url]
  (:body (client/get
          url
          {:accept "application/vnd.github.v3+json"
           :oauth-token github-token
           :as :json
           :query-params {:per_page 100}})))


;; https://docs.github.com/en/rest/reference/actions#list-repository-workflows
(defn workflows-url [owner repo]
  (str "https://api.github.com/repos/" owner "/" repo "/actions/workflows"))

;; https://docs.github.com/en/rest/reference/actions#list-workflow-runs
(defn workflow-run-url [owner repo run_id]
  (str (workflows-url owner repo) "/" run_id "/runs"))

(defn get-workflows [owner repo]
  (:workflows (get-url (workflows-url owner repo))))

(defn get-workflow-runs-by-id [owner repo run_id]
  (:workflow_runs (get-url (workflow-run-url owner repo run_id))))

(defn filter-by-name [name]
  (fn [item] (zero? (compare (:name item) name))))

(defn get-workflow-by-name [owner repo name]
  (first (->> (get-workflows owner repo)
              (filter (filter-by-name name)))))

(defn get-workflow-id-by-name [owner repo workflow-name]
  (:id (get-workflow-by-name owner repo workflow-name)))

;; TODO: add paging with page=1...x query param
(defn get-workflow-run-ids [owner repo workflow_id]
  (->> (get-workflow-runs-by-id owner repo workflow_id)
      ;;  (take 15)
       (map (fn [item] (:jobs_url item)))))

;; https://docs.github.com/en/rest/reference/actions#list-jobs-for-a-workflow-run
(defn get-jobs [job_url]
  (println "get-jobs" job_url)
  (:jobs (get-url job_url)))


src/ci_runtimes/core.clj
(ns ci-runtimes.core
  (:gen-class)
  (:require [clojure.pprint :as pprint]
            [clojure.instant :as instant]
            [ci-runtimes.github-api :as github-api]))

(defn line [n]
  (apply str (repeat n "#")))

(def max-line-length 15)

(defn line-for-duration [dur min-duration max-duration]
  (if (not= min-duration max-duration)
    (line (inc (* (- dur min-duration) (/ (dec max-line-length) (- max-duration min-duration)))))
    (line 1)))

(defn average [items]
  (if (pos? (count items))
    (/ (reduce + items) (count items))
    nil))

(defn diff-in-seconds [end start]
  (/ (- (.getTime end) (.getTime start)) 1000))

(defn diff-of-strings-in-seconds [end-str start-str]
  (when (and end-str start-str)
    (diff-in-seconds
     (instant/read-instant-date end-str)
     (instant/read-instant-date start-str))))

(defn date-to-iso-string [date]
  (.format (java.text.SimpleDateFormat. "yyyy-MM-dd") date))

(defn day-item-from-job [job] {:duration (diff-of-strings-in-seconds (:completed_at job) (:started_at job))
                               :date (date-to-iso-string (instant/read-instant-date (:started_at job)))})

(defn filter-success [job]
  (zero? (compare (:conclusion job) "success")))

(defn start-gather [owner repo workflow-name]
  (println "Get workflow" (str "\"" workflow-name "\"") "of" (str owner "/" repo) "...")
  (let [workflow_id (github-api/get-workflow-id-by-name owner repo workflow-name)]
    (println "Got workflow id:" workflow_id)
    (println "Get jobs of the workflow...")
    (let [job_urls (github-api/get-workflow-run-ids owner repo workflow_id)
          job (->> job_urls
                   (map github-api/get-jobs)
                   ;; TODO: specify which job to take from workflow run, now takes the first
                   (map first))
          day-items (->> job
                         (filter (complement nil?))  ;; CI job not started
                         (filter filter-success)
                         (map #(day-item-from-job %)))
          durations (->> day-items
                         (map #(:duration %))
                         (filter number?))
          average-duration (->> durations
                                (average)
                                (float))
          grouped-day-items (->> day-items
                                 (group-by :date) ;; -> {"2021-05-21" [{:date "2021-05-21" :duration 4}]}
                                 (map #(hash-map :date (key %)
                                                 :duration (->>
                                                            (val %)
                                                            (map :duration)
                                                            (average)
                                                            (float)
                                                            (Math/round))
                                                 :runs (->>
                                                        (val %)
                                                        (count)))))
          max-duration (->> grouped-day-items
                            (map #(:duration %))
                            (apply max))
          min-duration (->> grouped-day-items
                            (map #(:duration %))
                            (apply min))
          grouped-line-added (->> grouped-day-items
                                  (map #(assoc % :line (line-for-duration (:duration %) min-duration max-duration))))]
      ;; (println "Workflow run durations are" grouped-day-items)
      ;; (println "Workflow run durations are" grouped-line-added)
      (pprint/print-table (sort-by (fn [item] (:date item)) grouped-line-added))
      ;; TODO: humanize the durations
      (println)
      (println "Average duration of the workflow runs is" average-duration "seconds"))))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (if (== (count args) 3)
    (let [[owner repo workflow-name] args]
      (start-gather owner repo workflow-name))
    (do
      (println "Give params: owner, repo and workflow name")
      (println "Example: lein run colinhacks zod \"Fix, Format, Regenerate\""))))
